File: Dynamic Array/dynamicarray.cpp
#include <iostream>
#include <stdexcept>
#include "dynamicarray.hpp"

/* Constructor & Destructor */

DynamicArray::DynamicArray() {
  count = 0;
  capacity = 5;
  store = new int[capacity];
  fill();
}

DynamicArray::~DynamicArray() {
  delete [] store;
}

/* Public */

// O(1)
int DynamicArray::get(int index) {
  checkIndex(index);
  return store[index];
}

// O(1)
void DynamicArray::set(int index, int val) {
  checkIndex(index);
  store[index] = val;
}

// O(1)
void DynamicArray::push(int val) {
  if (count == capacity) resize();
  store[count] = val;
  count++;
}

// O(1)
int DynamicArray::pop() {
  checkIndex(count - 1);
  count--;
  store[count] = NULL;
  return store[count];
}

// O(n)
void DynamicArray::unshift(int val) {
  if (count == capacity) resize();
  for (int index = count - 1; index >= 0; index--) {
    store[index + 1] = store[index];
  }
  store[0] = val;
  count++;
}

// O(n)
int DynamicArray::shift() {
  checkIndex(0);
  int first = store[0];
  for (int index = 1; index < count; index++) {
    store[index - 1] = store[index];
  }
  count--;
  store[count] = NULL;
  return first;
}

void DynamicArray::print() {
  std::cout << "{ ";
  for (int index = 0; index < capacity; index++) {
    std::cout << store[index];
    if (index < capacity - 1) std::cout << ", ";
  }
  std::cout << " }\n" << std::endl;
}

/* Private */

void DynamicArray::checkIndex(int index) {
  if (isInvalid(index)) {
    throw std::out_of_range("Index out of bounds");
  }
}

bool DynamicArray::isInvalid(int index) {
  return (index < 0) || (index >= count);
}

void DynamicArray::resize() {
  int new_capacity = capacity * 2;
  int* new_store = new int[capacity];
  
  for (int index = 0; index < count; index++) {
    new_store[index] = store[index];
  }
  
  delete [] store;
  store = new_store;
  capacity = new_capacity;
  fill();
}

void DynamicArray::fill() {
  for (int index = count; index < capacity; index++) {
    store[index] = NULL;
  }
}

File: Dynamic Array/dynamic_array.rb
require_relative 'static_array'

class DynamicArray
  attr_reader :length, :capacity, :store

  def initialize
    @length = 0
    @capacity = 8
    @store = StaticArray.new(@capacity)
  end

  # O(1)
  def [](index)
    check_index_get(index)
    @store[index]
  end

  # O(1)
  def []=(index, value)
    check_index_set(index)
    @store[index] = value
  end

  # O(1)
  def pop
    last_value = self[@length - 1]
    @length -= 1
    last_value
  end

  # O(1) ammortized; O(n) worst case. Variable because of the possible
  def push(val)
    resize! if @length == @capacity
    self[@length] = val
    @length += 1
  end

  # O(n): has to shift over all the elements.
  def shift
    first_value = self[0]
    for index in 1...@length
      self[index - 1] = self[index]
    end
    @length -= 1
    first_value
  end

  # O(n): has to shift over all the elements.
  def unshift(val)
    resize! if @length == @capacity
    (@length - 1).downto(0) do |index|
      self[index + 1] = self[index]
    end
    self[0] = val
    @length += 1
  end

  protected
  def check_index_get(index)
    raise 'index out of bounds' if index >= @length
    raise 'index out of bounds' if index < 0
  end

  def check_index_set(index)
    raise 'index out of bounds' if index > @length
  end

  # O(n): has to copy over all the elements to the new store.
  def resize!
    new_capacity = @capacity * 2
    new_store = StaticArray.new(new_capacity)
    for index in 0...@length
      new_store[index] = self[index]
    end
    @capacity = new_capacity
    @store = new_store
  end
end